File: number-of-islands.js

var numIslands = function(grid) {
  
    let count = 0;
    for (let i = 0; i < grid.length; i++) {

        for( let j = 0; j < grid[i].length; j++) {
            if (grid[i][j] === '1') {

                count++;
                // Call the dfs function to mark all connected '1's as '0's
                dfs(grid, i, j);
            }
        }

    }
};

function dfs(grid, i, j) {

if(i<0 || i>=grid.length || j<0 || j>=grid[i].length || grid[i][j] === '0') {
    return;}

    grid[i][j] = '0'; // Mark the cell as visited
    dfs(grid, i + 1, j); // Down
    dfs(grid, i - 1, j); // Up
    dfs(grid, i, j + 1); // Right
    dfs(grid, i, j - 1); // Left

}

File: number-of-islands.py
def numIslands(self, grid):
        """
        :type grid: List[List[str]]
        :rtype: int
        """
        count =0
        for i in range(len(grid)):
            for j in range(len(grid[0])):
                if grid[i][j] == '1':
                    count += 1
                    self.dfs(grid,i,j)
        return count
def dfs(self, grid, i, j):
    if i<0 or i>=len(grid) or j<0 or j>=len(grid[0]) or grid[i][j] != '1':
        return
    grid[i][j] = '0'  # Mark the cell as visited
    self.dfs(grid, i+1, j)  # Down
    self.dfs(grid, i-1, j)  # Up
    self.dfs(grid, i, j+1)  # Right
    self.dfs(grid, i, j-1)  # Left